File: 128-Longest-Consecutive-Sequence/solution.ts
/**
 * Hash Table
 * Time is O(n) | Space O(n)
 * https://leetcode.com/problems/longest-consecutive-sequence/
 *
 * @param {number[]} nums
 * @return {number}
 */
function longestConsecutive(nums: number[]): number {
  const map: Map<number, number> = new Map();
  let maxLength = 0;

  for (let num of nums) {
    if (map.has(num)) continue;

    const leftLength = map.get(num - 1) || 0;
    const rightLength = map.get(num + 1) || 0;
    const totalLength = leftLength + rightLength + 1;

    map.set(num, totalLength);
    maxLength = Math.max(maxLength, totalLength);

    map.set(num - leftLength, totalLength);
    map.set(num + rightLength, totalLength);
  }

  return maxLength;
}

File: 128-Longest-Consecutive-Sequence/solution.go
package main

func longestConsecutive(nums []int) int {
	intSet := make(map[int]struct{}, len(nums))

	for _, num := range nums {
		if _, exists := intSet[num]; !exists {
			intSet[num] = struct{}{}
		}
	}

	longest := 0
	for num := range intSet {
		if _, exists := intSet[num-1]; !exists {
			length := 0
			for _, nextNumExists := intSet[num+length]; nextNumExists; _, nextNumExists = intSet[num+length] {
				length++
			}
			longest = max(longest, length)
		}
	}

	return longest
}
